sendrecv/gst-rust/src/main.rs

```diff
@@ -0,0 +1,480 @@
+    ServerConnecting = 1000,
+    ServerConnectionError,
+    ServerConnected,
+    ServerRegistering = 2000,
+    ServerRegisteringError,
+    ServerRegistered,
+    ServerClosed,
+    PeerConnecting = 3000,
+    PeerConnectionError,
+    PeerConnected,
+    PeerCallNegotiating = 4000,
+    PeerCallStarted,
+    PeerCallError,
+}
+#[derive(Debug)]
+enum Error {
+    GlibError(glib::Error),
+    GlibBoolError(glib::BoolError),
+    GstStateChangeError(gst::StateChangeError),
+}
+
+impl From<glib::Error> for Error {
+    fn from(error: glib::Error) -> Self {
+        Error::GlibError(error)
+    }
+}
+
+impl From<glib::BoolError> for Error {
+    fn from(error: glib::BoolError) -> Self {
+        Error::GlibBoolError(error)
+    }
+}
+
+impl From<gst::StateChangeError> for Error {
+    fn from(error: gst::StateChangeError) -> Self {
+        Error::GstStateChangeError(error)
+    }
+}
+
+const STUN_SERVER: &'static str = "stun-server=stun://stun.l.google.com:19302 ";
+const RTP_CAPS_OPUS: &'static str = "application/x-rtp,media=audio,encoding-name=OPUS,payload=";
+const RTP_CAPS_VP8: &'static str = "application/x-rtp,media=video,encoding-name=VP8,payload=";
+
+fn check_plugins() -> bool {
+    let needed = vec![
+        "opus",
+        "vpx",
+        "nice",
+        "webrtc",
+        "dtls",
+        "srtp",
+        "rtpmanager",
+        "videotestsrc",
+        "audiotestsrc",
+    ];
+    let registry = gst::Registry::get();
+    let mut ret = true;
+    for plugin_name in needed {
+        let plugin = registry.find_plugin(&plugin_name.to_string());
+        if plugin.is_none() {
+            println!("Required gstreamer plugin '{}' not found", plugin_name);
+            ret = false;
+        }
+    }
+    return ret;
+}
+
+fn setup_call(app_control: &Arc<Mutex<AppControl>>) -> AppState {
+    let mut app_control = app_control.lock().unwrap();
+    app_control.app_state = AppState::PeerConnecting;
+    println!(
+        "Setting up signalling server call with {}",
+        app_control.peer_id
+    );
+    app_control
+        .ws_sender
+        .send(format!("SESSION {}", app_control.peer_id))
+        .unwrap();
+    return AppState::PeerConnecting;
+}
+
+fn register_with_server(app_control: &Arc<Mutex<AppControl>>) -> AppState {
+    let mut app_control = app_control.lock().unwrap();
+    app_control.app_state = AppState::ServerRegistering;
+    let our_id = rand::thread_rng().gen_range(10, 10_000);
+    println!("Registering id {} with server", our_id);
+    app_control
+        .ws_sender
+        .send(format!("HELLO {}", our_id))
+        .unwrap();
+    return AppState::ServerRegistering;
+}
+
+fn send_sdp_offer(
+    app_control: &Arc<Mutex<AppControl>>,
+    offer: gstreamer_webrtc::WebRTCSessionDescription,
+) {
+    let app_control = app_control.lock().unwrap();
+    if app_control.app_state < AppState::PeerCallNegotiating {
+        // TODO signal and cleanup
+        panic!("Can't send offer, not in call");
```

Comment:
In the context of GStreamer what I usually do for reporting errors from somewhere in other threads to the main event loop is by posting a message on the bus. This could also transfer some Rust error type around.

See e.g. https://github.com/sdroege/gstreamer-rs/blob/85ca3b94a34d0236314aac900d5926c81ba73d99/examples/src/bin/decodebin.rs#L131-L139 and https://github.com/sdroege/gstreamer-rs/blob/85ca3b94a34d0236314aac900d5926c81ba73d99/examples/src/bin/decodebin.rs#L167-L185